src/Windows/Markdown/Editor/Impl/Preview/Code/RCodeBlock.cs

```diff
@@ -0,0 +1,128 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Common.Core;
+using Microsoft.Languages.Core.Tokens;
+using Microsoft.R.Core.Tokens;
+using Microsoft.R.Host.Client;
+using static System.FormattableString;
+
+namespace Microsoft.Markdown.Editor.Preview.Code {
+    internal sealed class RCodeBlock {
+        private StringBuilder _output;
+        private StringBuilder _errors;
+
+        public int BlockNumber { get; }
+        public string Text { get; }
+        public int Hash { get; }
+        public string Arguments { get; }
+        public string Result { get; set; }
+        public CodeBlockState State { get; set; } = CodeBlockState.Created;
+        public bool Eval { get; private set; } = true;
+        public bool DisplayErrors { get; private set; } = true;
+        public bool DisplayWarnings { get; private set; } = true;
+        public bool EchoContent { get; private set; } = true;
+
+        public string HtmlElementId => Invariant($"rcode_{BlockNumber}_{Hash}");
+
+        public RCodeBlock(int number, string text, string arguments = null) {
+            BlockNumber = number;
+            Text = text;
+            Arguments = arguments ?? string.Empty;
+            Hash = Text.GetHashCode() + Arguments.GetHashCode();
+            ExtractOptions(arguments);
+        }
+
+        public async Task<string> EvaluateAsync(IRSession session, RSessionCallback callback, CancellationToken ct) {
```

Comment:
Well it renders them into HTML document... :-) I.e evaluates, then transfers result to HTML doc according to the block id which matched HTMLElement id. It could be technically split into evaluator and renderer parts perhaps.